File: 3_chatgpt_sol2.cpp
#include <bits/stdc++.h>
using namespace std;

vector<pair<int, int>> primeFactors;
int factorCount;
vector<int> stateMultiplier;
vector<vector<int>> stateValues;
vector<vector<int>> mergedState;
unordered_map<long long, bool> dp;
int totalStates;

int encodeState(const vector<int> &exponents) {
    int id = 0;

    for (int i = 0; i < factorCount; i++) {
        id += exponents[i] * stateMultiplier[i];
    }

    return id;
}

bool isCursed(int firstState, int secondState) {

    long long key = ((long long) firstState << 32) | secondState;

    if (dp.count(key))
        return dp[key];

    int nextState = mergedState[firstState][secondState];

    if (nextState == totalStates - 1)
        return dp[key] = true;

    if (nextState == secondState)
        return dp[key] = false;

    return dp[key] = isCursed(secondState, nextState);
}

int main() {

    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int M;
    cin >> M;

    int temp = M;

    for (int p = 2; p * p <= temp; p++) {

        if (temp % p != 0)
            continue;

        int exponent = 0;

        while (temp % p == 0) {
            temp /= p;
            exponent++;
        }

        primeFactors.push_back({p, exponent});
    }

    if (temp > 1)
        primeFactors.push_back({temp, 1});

    factorCount = primeFactors.size();

    stateMultiplier.resize(factorCount);

    totalStates = 1;

    for (int i = 0; i < factorCount; i++) {
        stateMultiplier[i] = totalStates;
        totalStates *= (primeFactors[i].second + 1);
    }

    stateValues.assign(totalStates, vector<int>(factorCount));

    for (int state = 0; state < totalStates; state++) {

        int value = state;

        for (int i = 0; i < factorCount; i++) {

            int limit = primeFactors[i].second + 1;

            stateValues[state][i] = value % limit;
            value /= limit;
        }
    }

    mergedState.assign(totalStates, vector<int>(totalStates));

    for (int stateA = 0; stateA < totalStates; stateA++) {

        for (int stateB = 0; stateB < totalStates; stateB++) {

            vector<int> combined(factorCount);

            for (int i = 0; i < factorCount; i++) {

                combined[i] = min(
                    primeFactors[i].second,
                    stateValues[stateA][i] + stateValues[stateB][i]
                );
            }

            mergedState[stateA][stateB] = encodeState(combined);
        }
    }

    vector<int> numberState(M);

    for (int number = 1; number < M; number++) {

        int value = number;
        vector<int> exponents(factorCount, 0);

        for (int i = 0; i < factorCount; i++) {

            int prime = primeFactors[i].first;

            while (value % prime == 0) {
                exponents[i]++;
                value /= prime;
            }
        }

        numberState[number] = encodeState(exponents);
    }

    long long answer = 0;

    for (int x = 1; x < M; x++) {

        for (int y = 1; y < M; y++) {

            if (isCursed(numberState[x], numberState[y]))
                answer++;
        }
    }

    cout << answer << "\n";

    return 0;
}

File: 3_chatgpt_sol2.py
from functools import lru_cache

M = int(input())

prime_factors = []
temp = M
divisor = 2

while divisor * divisor <= temp:
    if temp % divisor == 0:
        exponent = 0
        while temp % divisor == 0:
            temp //= divisor
            exponent += 1
        prime_factors.append((divisor, exponent))
    divisor += 1

if temp > 1:
    prime_factors.append((temp, 1))

factor_count = len(prime_factors)

state_multiplier = []
total_states = 1

for _, exponent in prime_factors:
    state_multiplier.append(total_states)
    total_states *= (exponent + 1)

state_values = [[0] * factor_count for _ in range(total_states)]

for state in range(total_states):
    value = state
    for i, (_, exponent) in enumerate(prime_factors):
        state_values[state][i] = value % (exponent + 1)
        value //= (exponent + 1)


def encode_state(exponents):
    state = 0
    for i in range(factor_count):
        state += exponents[i] * state_multiplier[i]
    return state


merged_state = [[0] * total_states for _ in range(total_states)]

for state_a in range(total_states):
    for state_b in range(total_states):
        combined = []

        for i in range(factor_count):
            combined.append(
                min(
                    prime_factors[i][1],
                    state_values[state_a][i] + state_values[state_b][i]
                )
            )

        merged_state[state_a][state_b] = encode_state(combined)

number_state = [0] * M

for number in range(1, M):
    value = number
    exponents = [0] * factor_count

    for i, (prime, _) in enumerate(prime_factors):
        while value % prime == 0:
            exponents[i] += 1
            value //= prime

    number_state[number] = encode_state(exponents)


@lru_cache(None)
def is_cursed(first_state, second_state):
    next_state = merged_state[first_state][second_state]

    if next_state == total_states - 1:
        return True

    if next_state == second_state:
        return False

    return is_cursed(second_state, next_state)


answer = 0

for x in range(1, M):
    for y in range(1, M):
        if is_cursed(number_state[x], number_state[y]):
            answer += 1

print(answer)
